docs/src/adr/006rfc-unit-testing.md

```diff
@@ -0,0 +1,300 @@
+We were inspired by the design [Scalatest], [Scalacheck], [JML], and [QUIC
+testing] with Ivy. In comparison to pure property-based testing, we have to
+decompose the test in three parts:
+
+ - preparing the states (like in `Init`),
+ - executing the action (like a single instance of `Next`),
+ - testing the next states (like an invariant).
+
+### 3.3. Testing executions
+
+Engineers often like to test a particular set of executions to support their
+intuition, or to communicate an example to their peers. Sometimes, it is useful
+to isolate a set of executions to make continuous integration break, until the
+protocol is fixed. Needless to say, TLA+ tools have poor support for this,
+though they have all capabilities to produce such tests.
+
+Similar to testing an action in isolation, we propose an interface for testing
+a restricted set of executions as follows:
+
+```tla
+{{#include ../../../test/tla/ChangRobertsTyped_Test.tla:111:124}}
+```
+
+The test `TestExec_n0_n1` is similar to `TestAction_n0` in many aspects.  It
+starts by initializing the state with the predicate `Prepare_n0` and it expects
+a final state to satisfy the predicate `Assert_noWinner`. There is an important
+difference between the variables in `Assert_n0` and `Assert_noWinner`:
+
+ - Unprimed variables in `Assert_n0` refer to a state before firing an action,
+    whereas primed variables in `Assert_n0` refer to a state after firing
+    the action.
+
+ - Unprimed variables in `Assert_noWinner` refer to a state before firing
+   an *execution*, whereas primed variables in `Assert_noWinner` refer to
+   a final state of the execution.
+
+The annotation `@invariant("Correctness")` tell the tool to check the predicate
+`Correctness` in all the states of an execution.
+
+We should be able to run this test via:
+
+```sh
+apalache test --include=TestExec_n0_n1 ChangRobertsTyped_Test.tla
+```
+
+If the test is violated, a counterexample should be produced in the file
+`counterexample.tla`.
+
+
+## 4. Using tests for producing quick examples
+
+It is often nice to see examples of test inputs that pass the
+test. Apalache has all the ingredients to do that that. We should be able
+to run a command like that:
+
+```sh
+apalache example --include=TestAction_n0 ChangRobertsTyped_Test.tla
+```
+
+The above call would produce `example.tla`, a TLA+ description of two states
+that satisfy the test. This is similar to `counterexample.tla`, which is
+produced when an error is found.
+
+In a similar way we should be able to produce an example of an execution:
+
+```sh
+apalache example --include=TestExec_n0_n1 ChangRobertsTyped_Test.tla
+```
+
+
+[Unit testing]: https://en.wikipedia.org/wiki/Unit_testing
+[Property-based testing]: https://en.wikipedia.org/wiki/QuickCheck
+[TLA+ examples]: https://github.com/tlaplus/Examples/
+[LCR]: https://github.com/tlaplus/Examples/tree/master/specifications/chang_roberts
+[ChangRobertsTyped_Test.tla]: ../../../test/tla/ChangRobertsTyped_Test.tla
```

Comment:
I think this should be a link to the file hosted by github, otherwise the link will be broken when this is served from the generated site.

Reply:
Good point. I will have to update the link, once this PR is merged.